policy-controller/k8s/status/src/index.rs

```diff
@@ -0,0 +1,240 @@
+    client: k8s::Client,
+    updates: UnboundedReceiver<Update>,
+}
+
+pub struct Index {
+    updates: UnboundedSender<Update>,
+
+    http_routes: HashMap<ResourceId, Vec<ParentReference>>,
+    servers: HashSet<ResourceId>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Update {
+    pub id: ResourceId,
+    pub patch: k8s::Patch<serde_json::Value>,
+}
+
+impl Controller {
+    pub fn new(client: k8s::Client, updates: UnboundedReceiver<Update>) -> Self {
+        Self { client, updates }
+    }
+
+    pub async fn process_updates(mut self) {
+        let patch_params = k8s::PatchParams::apply("policy.linkerd.io");
+
+        // todo: If an update fails we should figure out a requeueing strategy
+        while let Some(Update { id, patch }) = self.updates.recv().await {
+            let api =
+                k8s::Api::<k8s::policy::HttpRoute>::namespaced(self.client.clone(), &id.namespace);
+
+            // todo: Do we need to consider a timeout here?
+            if let Err(error) = api.patch_status(&id.name, &patch_params, &patch).await {
+                tracing::error!(namespace = %id.namespace, name = %id.name, %error, "Failed to patch HTTPRoute");
+            }
+        }
+    }
+}
+
+impl Index {
+    pub fn shared(updates: UnboundedSender<Update>) -> SharedIndex {
+        Arc::new(RwLock::new(Self {
+            updates,
+            http_routes: HashMap::new(),
+            servers: HashSet::new(),
+        }))
+    }
+
+    // If the route is new or its parents have changed, return true so that a
+    // patch is generated; otherwise return false.
+    fn update_http_route(&mut self, id: ResourceId, parents: Vec<ParentReference>) -> bool {
+        match self.http_routes.entry(id) {
+            Entry::Vacant(entry) => {
+                entry.insert(parents);
+            }
+            Entry::Occupied(mut entry) => {
+                if *entry.get() == parents {
+                    return false;
+                }
+                entry.insert(parents);
+            }
+        }
+        true
+    }
+
+    fn make_http_route_patch(
+        &self,
+        id: &ResourceId,
+        parents: &[ParentReference],
+    ) -> k8s::Patch<serde_json::Value> {
+        let accepting_servers: Vec<ResourceId> = self
+            .servers
+            .iter()
+            .filter(|server| parents.iter().any(|parent| matches!(parent, ParentReference::Server(parent_id) if parent_id == *server)))
+            .cloned()
+            .collect();
+        let parent_statuses = parents
+            .iter()
+            .map(|parent| {
+                let ParentReference::Server(parent_reference_id) = parent;
+
+                #[cfg(not(test))]
+                let timestamp = Utc::now();
+                #[cfg(test)]
+                let timestamp = chrono::DateTime::<chrono::Utc>::MIN_UTC;
+
+                // Is this parent in the list of parents which accept
+                // the route?
+                let accepted = accepting_servers
```

Comment:
We build `accepting_servers` which is a Vec of all the servers which match a reference in the parent list.  Then for each parent in the parent list, we check to see if it matches something in the accepting_servers.  I think we can skip a step and do this more directly as:

```
let accepted = self.servers.iter().any(|server| server == parent_reference_id);
```

Reply:
Oh good call!